security/http/codegen.go

```diff
@@ -52,163 +73,221 @@ type SchemeData struct {
+		})
+	}
+}
+
+// getSchemesForEndpoint collects the security schemes for the given endpoint.
+func getSchemesForEndpoint(e *httpdesign.EndpointExpr) []*SchemeData {
+	payload := e.MethodExpr.Payload
+	var schemes []*SchemeData
+	for _, req := range design.Requirements(e.Service.Name(), e.Name()) {
+		for _, s := range req.Schemes {
+			scheme := *s
+			switch scheme.Kind {
+			case design.BasicAuthKind:
+				if goadesign.IsObject(payload.Type) {
+					userAtt, user := findSecurityField(payload, "security:username")
+					passAtt, pass := findSecurityField(payload, "security:password")
+					schemes = append(schemes, &SchemeData{
+						UsernameField:   user,
+						UsernamePointer: payload.IsPrimitivePointer(userAtt, true),
+						PasswordField:   pass,
+						PasswordPointer: payload.IsPrimitivePointer(passAtt, true),
+						Scheme:          &scheme,
+					})
+				}
+
+			case design.APIKeyKind:
+				if goadesign.IsObject(payload.Type) {
+					keyAtt, key := findSecurityField(payload, "security:apikey:"+scheme.SchemeName)
+					if key == "" {
+						continue
+					}
+					name := findMappedName(e.AllParams(), keyAtt)
+					if name != "" {
+						scheme.In = "query"
+					} else {
+						name = findMappedName(e.MappedHeaders(), keyAtt)
+						scheme.In = "header"
+					}
+					scheme.Name = name
+					schemes = append(schemes, &SchemeData{
+						CredField:   key,
+						CredPointer: payload.IsPrimitivePointer(keyAtt, true),
+						Scheme:      &scheme,
+					})
+				}
+
+			case design.JWTKind:
+				if goadesign.IsObject(payload.Type) {
+					keyAtt, key := findSecurityField(payload, "security:token")
+					if key == "" {
+						continue
+					}
+					schemes = append(schemes, &SchemeData{
+						CredField:   key,
+						CredPointer: payload.IsPrimitivePointer(keyAtt, true),
+						Scheme:      &scheme,
+					})
+				}
+
+			case design.OAuth2Kind:
+				if goadesign.IsObject(payload.Type) {
+					keyAtt, key := findSecurityField(payload, "security:accesstoken")
+					if key == "" {
+						continue
+					}
+					name := findMappedName(e.AllParams(), keyAtt)
+					if name != "" {
+						scheme.In = "query"
+					} else {
+						name = findMappedName(e.MappedHeaders(), keyAtt)
+						scheme.In = "header"
+					}
+					if name != "" {
+						scheme.Name = name
+					}
+					schemes = append(schemes, &SchemeData{
+						CredField:   key,
+						CredPointer: payload.IsPrimitivePointer(keyAtt, true),
+						Scheme:      &scheme,
+					})
+				}
+
+			default:
+				panic(fmt.Sprintf("unknown kind %#v", scheme.Kind)) // bug
 			}
-			f.SectionTemplates = append(f.SectionTemplates, &codegen.SectionTemplate{
-				Name:    "secure-request-decoder",
-				Source:  authDecoderT,
-				Data:    secureData,
-				FuncMap: codegen.TemplateFuncs(),
-			})
 		}
 	}
+	return schemes
+}
+
+// findMappedName returns the mapped name for the given attribute name.
+func findMappedName(p *goadesign.MappedAttributeExpr, keyAtt string) string {
```

Comment:
I wonder if we shouldn't add that as a method of the `MappedAttributeExpr` struct instead? This is a legitimate case where one might not know whether the key exists a-priori which isn't a case that the struct was designed for originally. Maybe the struct could have a `FindKey` method that accepts the name of the key and returns the name of the corresponding element and `true` if there is one otherwise empty string and `false`?
Also the test for `obj` being `nil` is not needed as a mapped attribute type is always object (`NewMappedAttribute` panics otherwise).

Reply:
Added the function like you suggested here - https://github.com/goadesign/goa/pull/1478/commits/d6a66808843b7eb4431cf9aa33e932cc993bf3a5